pkg/controller/tide/tide.go

```diff
@@ -0,0 +1,730 @@
+				logger.Info("can not release online node to offline", "node", onlineNodesInfo.node.Name)
+				return nil
+			}
+			pod.Spec.NodeName = nodeName
+			clusterSnapshot.AddPod(pod, nodeName)
+		}
+	}
+	node := t.changeNodeToOffline(nodeInfo.Node(), tideNodePool)
+	if _, err := t.client.KubeClient.CoreV1().Nodes().Update(ctx, node, metav1.UpdateOptions{}); err != nil {
+		return err
+	}
+	logger.Info("release online node success", "node", node.Name)
+	return nil
+}
+
+func (t *Tide) changeNodeToOnline(node *corev1.Node, pool NodePoolWrapper) *corev1.Node {
+	found := false
+	for i := range node.Spec.Taints {
+		if node.Spec.Taints[i].Key == pool.GetEvictOnlinePodTaint().Key || node.Spec.Taints[i].Key == pool.GetEvictOfflinePodTaint().Key {
+			node.Spec.Taints[i].Key = pool.GetEvictOfflinePodTaint().Key
+			node.Spec.Taints[i].Value = pool.GetEvictOfflinePodTaint().Value
+			node.Spec.Taints[i].Effect = corev1.TaintEffect(pool.GetEvictOfflinePodTaint().Effect)
+			found = true
+		}
+	}
+	if !found {
+		node.Spec.Taints = append(node.Spec.Taints, corev1.Taint{
+			Key:    pool.GetEvictOfflinePodTaint().Key,
+			Value:  pool.GetEvictOnlinePodTaint().Value,
+			Effect: corev1.TaintEffect(pool.GetEvictOnlinePodTaint().Effect),
+		})
+	}
+	delete(node.Labels, pool.GetOfflineLabel().Key)
+	node.Labels[pool.GetOnlineLabel().Key] = pool.GetOnlineLabel().Value
+	return node
+}
+
+func (t *Tide) changeNodeToOffline(node *corev1.Node, pool NodePoolWrapper) *corev1.Node {
+	found := false
+	for i := range node.Spec.Taints {
+		if node.Spec.Taints[i].Key == pool.GetEvictOnlinePodTaint().Key || node.Spec.Taints[i].Key == pool.GetEvictOfflinePodTaint().Key {
+			node.Spec.Taints[i].Key = pool.GetEvictOnlinePodTaint().Key
+			node.Spec.Taints[i].Value = pool.GetEvictOnlinePodTaint().Value
+			node.Spec.Taints[i].Effect = corev1.TaintEffect(pool.GetEvictOnlinePodTaint().Effect)
+			found = true
+		}
+	}
+	if !found {
+		node.Spec.Taints = append(node.Spec.Taints, corev1.Taint{
+			Key:    pool.GetEvictOfflinePodTaint().Key,
+			Value:  pool.GetEvictOnlinePodTaint().Value,
+			Effect: corev1.TaintEffect(pool.GetEvictOnlinePodTaint().Effect),
+		})
+	}
+	delete(node.Labels, pool.GetOnlineLabel().Key)
+	node.Labels[pool.GetOfflineLabel().Key] = pool.GetOfflineLabel().Value
+	return node
+}
+
+func getNodeUsageWithSelector(
+	nodes simulator.ClusterSnapshot,
+	resourceNames []corev1.ResourceName,
+	selector labels.Selector,
+) ([]NodeUsage, error) {
+	var nodeUsageList []NodeUsage
+	nodeInfos, err := nodes.NodeInfos().List()
+	if err != nil {
+		return nil, err
+	}
+	for _, node := range nodeInfos {
+		var pods []*corev1.Pod
+		for i := range node.Pods {
+			pods = append(pods, node.Pods[i].Pod)
+		}
+		if !selector.Matches(labels.Set(node.Node().GetLabels())) {
+			continue
+		}
+		nodeUsageList = append(nodeUsageList, NodeUsage{
+			node:    node.Node(),
+			usage:   nodeutil.NodeUtilization(pods, resourceNames),
+			allPods: pods,
+		})
+	}
+	// 节点按照 使用率排序
```

Comment:
please change to english version for comments

Reply:
the same for other comments

Reply:
done